libraries/standard/mqtt/src/mqtt.c

```diff
@@ -714,6 +758,81 @@ static MQTTStatus_t handleIncomingAck( MQTTContext_t * const pContext,
             break;
     }
 
+    if( ( status == MQTTSuccess ) && ( callAppCallback == true ) )
+    {
+        appCallback( pContext, pIncomingPacket, packetIdentifier, NULL );
+    }
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+static MQTTStatus_t receiveSingleIteration( MQTTContext_t * pContext,
+                                            uint32_t remainingTimeMs,
+                                            bool manageKeepAlive )
+{
+    MQTTStatus_t status = MQTTSuccess;
+    MQTTPacketInfo_t incomingPacket;
+
+    assert( pContext != NULL );
+
+    status = MQTT_GetIncomingPacketTypeAndLength( pContext->transportInterface.recv,
+                                                  pContext->transportInterface.networkContext,
+                                                  &incomingPacket );
+
+    if( status == MQTTNoDataAvailable )
+    {
+        if( manageKeepAlive == true )
+        {
+            /* Assign status so an error can be bubbled up to application,
+             * but reset it on success. */
+            status = handleKeepAlive( pContext );
+        }
+
+        if( status == MQTTSuccess )
+        {
+            /* Reset the status to indicate that we should not try to read
+             * a packet from the transport interface. */
+            status = MQTTNoDataAvailable;
+        }
+    }
+    else if( status != MQTTSuccess )
+    {
+        LogError( ( "Receiving incoming packet length failed. Status=%d",
+                    status ) );
+    }
+    else
+    {
+        /* Receive packet. Remaining time is recalculated before calling this
+         * function. */
+        status = receivePacket( pContext, incomingPacket, remainingTimeMs );
+    }
+
+    /* Handle received packet. If no data was read then this will not execute. */
+    if( status == MQTTSuccess )
+    {
+        incomingPacket.pRemainingData = pContext->networkBuffer.pBuffer;
+
+        /* PUBLISH packets allow flags in the lower four bits. For other
+         * packet types, they are reserved. */
+        if( ( incomingPacket.type & 0xF0U ) == MQTT_PACKET_TYPE_PUBLISH )
+        {
+            status = handleIncomingPublish( pContext, &incomingPacket );
+        }
+        else
+        {
+            status = handleIncomingAck( pContext, &incomingPacket, manageKeepAlive );
+        }
+    }
+
+    if( status == MQTTNoDataAvailable )
```

Comment:
If `MQTTNoDataAvailable` is not passed to the user, then should it be a public-facing `MQTTStatus_t` type?

Reply:
Yes, it's one of the possible return values of `MQTT_GetIncomingPacketTypeAndLength` in the lightweight API